File: 7 StacksandQues/Brackets/brackets.java
import java.util.Stack;
class Solution {
  public int solution(String S) {
    if (S.length() == 0)
      return 1;

    Stack<Character> stk = new Stack<>();

    for (char c : S.toCharArray()) {
      if (c == '(' || c == '{' || c == '[') {
        stk.push(c);
      }
      if (stk.isEmpty())
        return 0;

      if (c == ')') {
        if (stk.pop() != '(') return 0;
      }
      if (c == '}') {
        if (stk.pop() != '{') return 0;
      }
      if (c == ']') {
        if (stk.pop() != '[') return 0;
      }
    }
    if (!stk.isEmpty()) {
      return 0;
    } else {
      return 1;
    }
  }
}

File: 7 StacksandQues/Brackets/brackets.cpp
#include <stack>
int solution(string &S) {
  if (S.size() == 0) return 1;  // if input empty
  if (S.size() % 2) return 0;   // if odd input cant be properly nested

  stack<char> stk;

  for (auto &s : S)
    if (s == '(' or s == '[' or s == '{')
      stk.push(s);
    else if (stk.size() == 0)
      return false;
    else if (s == ')' and stk.top() != '(')
      return false;
    else if (s == ']' and stk.top() != '[')
      return false;
    else if (s == '}' and stk.top() != '{')
      return false;
    else
      stk.pop();

  return stk.size() == 0;
}
